File: 2018/08/201808.py
import sys
from typing import Iterator,List,Any
from dataclasses import dataclass

class Node:
    def __init__(self) -> None:
        self.meta: List[int] = []
        self.children: List[Node] = []

def read_tree(ns: Iterator[int]) -> Node:
    node = Node()
    n_nodes = next(ns)
    n_meta = next(ns)

    for i in range(n_nodes):
        child = read_tree(ns)
        node.children.append(child)

    for i in range(n_meta):
        node.meta.append(next(ns))

    return node

def sum_meta(node: Node) -> int:
    return sum(node.meta) + sum(sum_meta(n) for n in node.children)

def sum_code(node: Node) -> int:
    return (sum(node.meta)
            if len(node.children) == 0
            else sum(sum_code(node.children[n-1])
                for n in node.meta if n >=1 and n <= len(node.children)))

def main() -> None:
    data = (int(s) for s in sys.stdin.read().split())
    tree = read_tree(data)

    ans1 = sum_meta(tree)
    print(ans1)

    ans2 = sum_code(tree)
    print(ans2)

if __name__ == '__main__':
    main()

File: 2018/08/201808.jl


struct Node
    meta::Vector{Int32}
    kids::Vector{Node}
end

Node() = Node([],[])

function read_tree(ns::Vector{Int32})::Node
    node = Node()
    n_nodes = popfirst!(ns)
    n_meta = popfirst!(ns)

    for i in 1:n_nodes
        child = read_tree(ns)
        push!(node.kids, child)
    end

    for i in 1:n_meta
        push!(node.meta, popfirst!(ns))
    end

    node
end

function sum_meta(node::Node)::Int32
    sum(node.meta) + mapfoldl(sum_meta, +, node.kids; init=Int32(0))
end

function sum_code(node::Node)::Int32
    if length(node.kids) == 0
        sum(node.meta)
    else
        s = Int32(0)
        for ni in node.meta
            if ni > 0 && ni <=length(node.kids)
                s += sum_code(node.kids[ni])
            end
        end
        s
    end
end

function main()
    data = [parse(Int32, s) for s in split(read(open("input.txt"),String))]
    tree = read_tree(data)

    ans1 = sum_meta(tree)
    println("PART1: ", ans1)

    ans2 = sum_code(tree)
    println("PART2: ", ans2)
end

main()



